platform/linux-generic/odp_packet_io.c

```diff
@@ -2851,11 +2941,53 @@ uint64_t odp_pktin_wait_time(uint64_t nsec)
 	return (nsec / (1000)) + 1;
 }
 
+static int allocate_tx_compl_events(const pktio_entry_t *entry, const odp_packet_t packets[],
+				    int num, tx_compl_info_t *tx_compl_info, uint16_t *num_tx_cevs)
+{
+	odp_packet_t pkt;
+	odp_packet_hdr_t *pkt_hdr;
+	odp_event_t ev;
+	int i;
+
+	for (i = 0; i < num; i++) {
+		pkt = packets[i];
+		pkt_hdr = packet_hdr(pkt);
+
+		if (odp_unlikely(pkt_hdr->p.flags.tx_compl)) {
+			ev = _odp_pktio_allocate_tx_compl_event(entry, odp_packet_user_ptr(pkt));
```

Comment:
Single odp_buffer_alloc_multi() is much more efficient than odp_buffer_alloc() in a loop.

So, first packets would be looped through to find those (couple of) packets that need tx_compl. Then alloc multi call, which determines up to which index we can send.

It's possible that we'll have other per packet checks done at top level send call, so this for-each-packet loop could be raised up into the top level.

Reply:
As you mentioned, there are likely just a few completion event requiring packets so probably not much difference allocating the events for those few already in the loop.

Reply:
Using multi allocation in v5.